src/hotspot/os/linux/cgroupSubsystem_linux.cpp

```diff
@@ -301,6 +327,23 @@ bool CgroupSubsystemFactory::determine_type(CgroupInfo* cg_infos,
     // block in the hybrid case.
     //
     if (is_cgroupsV2 && sscanf(p, "%*d %*d %*d:%*d %*s %s %*[^-]- %s %*s %*s", tmp_mount_point, tmp_fs_type) == 2) {
+      // On some systems duplicate controllers get mounted in addition to
+      // the main cgroup controllers most likely under /sys/fs/cgroup. In that
+      // case pick the first one under /sys/fs/cgroup and discard others.
+      if (cgroupv2_mount_point_found && strcmp("cgroup2", tmp_fs_type) == 0) {
+        if (strstr(cg_infos[0]._mount_path, "/sys/fs/cgroup") != cg_infos[0]._mount_path &&
+            strstr(tmp_mount_point, "/sys/fs/cgroup") == tmp_mount_point) {
+          log_debug(os, container)("Duplicate cgroupv2 controllers detected. Picking %s, skipping %s.",
+                                   tmp_mount_point, cg_infos[0]._mount_path);
+          for (int i = 0; i < CG_INFO_LENGTH; i++) {
+            os::free(cg_infos[i]._mount_path);
+            cg_infos[i]._mount_path = os::strdup(tmp_mount_point);
+          }
+        } else {
+          log_debug(os, container)("Duplicate cgroupv2 controllers detected. Picking %s, skipping %s.",
+                                   cg_infos[0]._mount_path, tmp_mount_point);
+        }
+      }
```

Comment:
Couldn't we simplify this? For example by changing:

```
      if (!cgroupv2_mount_point_found && strcmp("cgroup2", tmp_fs_type) == 0) {
        cgroupv2_mount_point_found = true;
        any_cgroup_mounts_found = true;
        for (int i = 0; i < CG_INFO_LENGTH; i++) {
          assert(cg_infos[i]._mount_path == NULL, "_mount_path memory stomping");
          cg_infos[i]._mount_path = os::strdup(tmp_mount_point);
        }
```

to something like this:

```
      if (strcmp("cgroup2", tmp_fs_type) == 0) {
        cgroupv2_mount_point_found = true;
        any_cgroup_mounts_found = true;
        for (int i = 0; i < CG_INFO_LENGTH; i++) {
          set_controller_paths(cg_infos, i, "(cg2, unified)", tmpmount, tmproot);
        }
```

It would entail modifying the `sscanf` to something like `sscanf(p, "%*d %*d %*d:%*d %s %s %*[^-]- %s %*s %*s", tmproot, tmpmount, tmp_fs_type) == 3`. We could also get rid of duplicate local var `tmp_mount_point`.

Reply:
done